fluentd/configs.d/openshift/output-operations.conf

```diff
@@ -1,6 +1,7 @@
-<match output_ops_tag journal.** system.var.log** mux.ops audit.log** %OCP_FLUENTD_TAGS%>
+<match ops_fluentd_tag journal.system** system.var.log** mux.ops audit.log** %OCP_FLUENTD_TAGS%>
```

Comment:
The tag name is being visible in the forwarded rsyslog output.
The QE filed bug [https://bugzilla.redhat.com/show_bug.cgi?id=1543761] complained that "The output_tag is also wrong".  I wanted to make it a little more user friendly.  Could there be a better/clearer keyword you could think of?

This is the new behaviour:
If use_record=true, operations logs would have systemd.u.SYSLOG_IDENTIFIER value.
Other messages from the containers have the tag "fluentd_tag".
```
Sample logs in /var/log/messages in the precise format:
3,17,Feb 15 14:52:25,ip-172-18-5-234.ec2.internal,rsyslogTestTag:, rsyslogTestMessage-20180215-145124
6,16,Feb 15 14:52:41,ip-172-18-5-234.ec2.internal,fluentd_tag:, namespace_name=logging, container_name=kibana, pod_name=logging-kibana-1-qr6pl, message=GET /testKibanaMessage-20180215-145225 404 4ms - 9.0B
```
If use_record=false, operations log has "ops_fluentd_tag" and container log has "fluentd_tag"

By default, use_record=false.  (I have to fix "use_record true, by default." in the git comment, which is wrong... :p)

Reply:
Changing the tags is going to make backporting anything very difficult and error prone.

Changing the tags also might cause problems for customers if they are already using `output_tag` and `output_ops_tag`.

```
$ logger -t newYear -p local0.notice happy_new_year

# cat /var/log/rsyslog.log | grep -i year
local3 info Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year [[<158>Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year]]

The output_tag is also wrong, it should be newYear in this case
```
The `-t` argument to logger is the `ident` field aka the `systemd.u.SYSLOG_IDENTIFIER` field, not the fluentd tag.  I don't think we should be sending the fluentd tag with the remote syslog protocol.

Reply:
Ok.  I was hoping to put some meaningful word like retagged_by_fluentd or something like that.  But it seems it's rather harmful to change it.
- I'm reverting the tag name change and put them back to output_tag and output_ops_tag.
- remote-syslog plugin is implemented to send the fluentd tag to rsyslogd by default.  (Or you could specify a record key to alter.)  I could change the default behaviour to send ident/systemd.u.SYSLOG_IDENTIFIER as a tag to rsyslogd for the operations logs.  How about the container logs?  I could not find ident/systemd.u.SYSLOG_IDENTIFIER in the container logs.   Do you have any suggestion which field could be used for the container logs' tag?

Reply:
> remote-syslog plugin is implemented to send the fluentd tag to rsyslogd by default

In which field of the output does it add the fluentd tag?  Is this some sort of spec/rfc, or is it just by convention?

For example: https://tools.ietf.org/html/rfc3164#section-5.3
What they refer to as the `TAG` field is really the ident/systemd.u.SYSLOG_IDENTIFIER

https://tools.ietf.org/html/rfc5424#section-6.2.5 - here it is called the APP-NAME field, but it is really the same as the ident/systemd.u.SYSLOG_IDENTIFIER

If it really is specific to the remote-syslog plugin, and users of the remote-syslog plugin expect the fluentd tag to be in that field, then
- we have to do that by default so as not to break existing applications
- users should not care, by default, what that value is
- if the remote-syslog plugin already allows you to change that field value, then we should still honor that

Given that, I don't see why this is wrong:
```
local3 info Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year [[<158>Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year]]
```
if the remote-syslog plugin is doing exactly what it is supposed to do in that case, it isn't "wrong".

Reply:
re: container logs - when reading from journald, container logs typically use `docker.service` (which will probably change when cri-o is used instead of docker).  When reading from json-file, correct, there is no ident/systemd.u.SYSLOG_IDENTIFIER field.  Does the remote-syslog plugin already have the ability to specify what to put in the tag/ident field if there is no appropriate value in the record?

Reply:
> In which field of the output does it add the fluentd tag?

This code assigns the fluentd tag to the tag field of the packet to be sent to rsyslogd.
https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/out_syslog_buffered.rb#L115

Regarding this question:
> Given that, I don't see why this is wrong:
> local3 info Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year [[<158>Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year]]

The test case in bz1543761 sets use_record=true.
```openshift_logging_fluentd_remote_syslog_use_record=true```
If it is true, the facility and the severity from this logger command are supposed to be respected.
```$ logger -t newYear -p local0.notice happy_new_year```
However, the log found in /var/log/rsyslog.log did not contain ```local0 notice``` but ```local3 info``` which pair comes from output-remote-syslog.conf [1].  Junqi claimed the value pair is different from the expected one.  So is the tag.  First, I thought the tag part is not a bug but a spec and it could be acceptable.  But rethinking of it, I had to agree losing ```-t newYear``` was not nice.  So, I added the code "if use_record is true, pass ident/systemd.u.SYSLOG_IDENTIFIER to rsyslogd."
```
# cat /var/log/rsyslog.log | grep -i year
local3 info Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year [[<158>Feb  9 03:21:58 logging-fluentd-pzgrk output_tag: happy_new_year]]
```
The facility, severity bug is caused here in the remote-syslog plugin code, where it tries to get the facility, severity values as follows.
https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/out_syslog_buffered.rb#L95-L97

In reality, record['facility'] and record['severity'] do not exist and some tweaks were needed to pass the values to rsyslogd.
https://github.com/nhosoi/origin-aggregated-logging/blob/bz1515715/fluentd/out_syslog_buffered.rb#L96-L103

[1]
```
<store>
@type syslog_buffered
remote_syslog 172.18.9.41
port 514
hostname logging-fluentd-####
facility local3
severity info
use_record false
</store>
```
> Does the remote-syslog plugin already have the ability to specify what to put in the tag/ident field if there is no appropriate value in the record?

This is the current tag retrieving code.
https://github.com/nhosoi/origin-aggregated-logging/blob/bz1515715/fluentd/out_syslog_buffered.rb#L120-L136

Reply:
> This code assigns the fluentd tag to the tag field of the packet to be sent to rsyslogd.
https://github.com/openshift/origin-aggregated-logging/blob/master/fluentd/out_syslog_buffered.rb#L115

- by default it uses the fluentd tag - this is the existing behavior and should not be changed
- user can specify tag_key to use a different value from the record, so is customizable already

> However, the log found in /var/log/rsyslog.log did not contain local0 notice but local3 info which pair comes from output-remote-syslog.conf [1]. Junqi claimed the value pair is different from the expected one.

That sounds like a real bug that we should fix.

> So is the tag.

I don't see this is a bug.  The tag is set correctly to the fluentd tag.  This is not affected by `use_record=true` or `false`.  If the user doesn't want the fluentd tag to be in this field, the user can specify `tag_key`

> First, I thought the tag part is not a bug but a spec and it could be acceptable. But rethinking of it, I had to agree losing -t newYear was not nice.

But that's the way the remote-syslog plugin works - it isn't a bug, it is a feature.  If this is not desirable, then either use `tag_key`, or QE can file a separate RFE to make the `-t` `ident` field preserved (somewhere?) by default.

> So, I added the code "if use_record is true, pass ident/systemd.u.SYSLOG_IDENTIFIER to rsyslogd."

I don't think that should be done.  From what I can see, the user can set `tag_key` if they want the `ident` value instead of the fluentd tag.

If QE insists that `ident` should be used here, they should file a separate bz/RFE.

Reply:
> But that's the way the remote-syslog plugin works - it isn't a bug, it is a feature. If this is not desirable, then either use tag_key, or QE can file a separate RFE to make the -t ident field preserved (somewhere?) by default.

Sounds good.  I have some question here... 

In filter-viaq-data-model.conf, ident and SYSLOG_IDENTIFIER are being removed as follows.  Could it be ok to leave and refer from remote-syslog plugin with ```tag_key ident``` or ```tag_key SYSLOG_IDENTIFIER```?  Unfortunately, only one tag_key is accepted, though.  Do we want to extend it to multiple?
```
  <formatter>
    tag "system.var.log**"
    remove_keys host,pid,ident
  </formatter>
  <formatter>
    tag "journal.system**" as well as "kubernetes.journal.container**"
    remove_keys ...,SYSLOG_IDENTIFIER,...
  </formatter>
```
Another question is in the current remote-syslog plugin gode, tag_key is referred as record[@tag_key] and there is no way to access ```record[syslogd][u][SYSLOG_IDENTIFIER]```.  Do you think we need some change to accept ```tag_key systemd.u.SYSLOG_IDENTIFIER``` and convert it to ```record[syslogd][u][SYSLOG_IDENTIFIER]```?  Or once we could make ident and SYSLOG_IDENTIFIER work as tag_key, we don't need this ```tag_key systemd.u.SYSLOG_IDENTIFIER``` support?

Reply:
> In filter-viaq-data-model.conf, ident and SYSLOG_IDENTIFIER are being removed as follows.

Right, because `ident` from /var/log/messages (which is deprecated) and `SYSLOG_IDENTIFIER` from the journal are converted to `systemd.u.SYSLOG_IDENTIFIER`.  Once that is done, we don't need the original `ident` and `SYSLOG_IDENTIFIER` fields any more.

Multiple tag keys would also be a good enhancement.

I guess the problem is that you cannot use a nested `tag_key`.  I think then the fix should be that you should be able to specify something like `tag_key systemd.u.SYSLOG_IDENTIFIER` which will use `record['systemd']['u']['SYSLOG_IDENTIFIER']`.  That would be a good enhancement to this plugin.   rewrite-tag-filter added something like this: https://github.com/fluent/fluent-plugin-rewrite-tag-filter#nested-attributes

Reply:
Ok. Thanks!